ndc-lambda-sdk/src/inference.ts

```diff
@@ -226,154 +237,209 @@ function getParallelDegreeFromJsDoc(functionSymbol: ts.Symbol, functionIsReadonl
-    return new Err([`Function types are not supported, but one was encountered in ${typePathToString(typePath)} (type: ${context.typeChecker.typeToString(tsType)})`]);
+    return new Err([`Function types are not supported, but one was encountered in ${schema.typePathToString(typePath)} (type: ${context.typeChecker.typeToString(tsType)})`]);
   }
 
   if (isMapType(tsType)) {
-    return new Err([`Map types are not supported, but one was encountered in ${typePathToString(typePath)} (type: ${context.typeChecker.typeToString(tsType)})`]);
+    return new Err([`Map types are not supported, but one was encountered in ${schema.typePathToString(typePath)} (type: ${context.typeChecker.typeToString(tsType)})`]);
   }
 
   const schemaTypeResult =
-    deriveSchemaTypeIfTsArrayType(tsType, typePath, context, recursionDepth)
+    deriveSchemaTypeIfTsUnknownOrAny(tsType, typePath, allowRelaxedTypes, context)
+    ?? deriveSchemaTypeIfTsTupleType(tsType, typePath, allowRelaxedTypes, context, recursionDepth)
+    ?? deriveSchemaTypeIfTsArrayType(tsType, typePath, allowRelaxedTypes, context, recursionDepth)
     ?? deriveSchemaTypeIfScalarType(tsType, context)
-    ?? deriveSchemaTypeIfNullableType(tsType, typePath, context, recursionDepth)
-    ?? deriveSchemaTypeIfObjectType(tsType, typePath, context, recursionDepth);
+    ?? deriveSchemaTypeIfNullableType(tsType, typePath, allowRelaxedTypes, context, recursionDepth)
+    ?? deriveSchemaTypeIfObjectType(tsType, typePath, allowRelaxedTypes, context, recursionDepth)
+    ?? rejectIfClassType(tsType, typePath, context) // This needs to be done after scalars, because JSONValue is a class
+    ?? deriveSchemaTypeIfTsIndexSignatureType(tsType, typePath, allowRelaxedTypes, context, recursionDepth) // This needs to be done after scalars and classes, etc because some of those types do have index signatures (eg. strings)
+    ?? deriveSchemaTypeIfTsUnionType(tsType, typePath, allowRelaxedTypes, context, recursionDepth); // This needs to be done after nullable types, since nullable types use unions, this catches everything else
 
   if (schemaTypeResult !== undefined)
     return schemaTypeResult;
 
-  // Types with index signatures: ie '[x: T]: Y'
-  if (context.typeChecker.getIndexInfosOfType(tsType).length > 0) {
-    return new Err([`Types with index signatures are not supported, but one was encountered in ${typePathToString(typePath)} (type: ${context.typeChecker.typeToString(tsType)})`]);
+  // We don't know how to deal with this type, so reject it with a generic error
+  return new Err([`Unable to derive an NDC type for ${schema.typePathToString(typePath)} (type: ${context.typeChecker.typeToString(tsType)}).`]);
+}
+
+function deriveRelaxedTypeOrError(typeName: string, typePath: schema.TypePathSegment[], mkError: () => string, allowRelaxedTypes: boolean, context: TypeDerivationContext): Result<schema.TypeReference, string[]> {
+  if (allowRelaxedTypes === false) {
+    return new Err([mkError()]);
   }
```

Comment:
Yes, it's used by the other functions that identify the appropriate types that can be turned into relaxed types (eg `unknown` or `any`).

Reply:
Ah I see. I feel like it's a bit of a smell to pass in a bool like this, since if you already know, then just don't call the function, but I don't mind if you want to keep it this way.

Reply:
Oh sorry, I may not have explained properly. This is _always_ called in the functions that identify types that may be relaxed types. This function is responsible for bailing out if relaxed types is disallowed.

For example, see the `deriveSchemaTypeIfTsTupleType`. Yeah, there's an `allowRelaxedTypes` check in there, but it's that's really just preventing un-needed computation if relaxed types is disabled (Haskell's laziness would have been nice here 🥲).